cns/restserver/ipampoolmonitor.go

```diff
@@ -0,0 +1,138 @@
+package restserver
+
+import (
+	"context"
+	"sync"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/logger"
+	"github.com/Azure/azure-container-networking/cns/requestcontroller"
+	nnc "github.com/Azure/azure-container-networking/nodenetworkconfig/api/v1alpha"
+)
+
+var (
+	increasePoolSize = 1
+	decreasePoolSize = -1
+	doNothing        = 0
+)
+
+type IPAMPoolMonitor interface {
+	Start() error
+	UpdatePoolLimitsTransacted(batchSize int, requestThreshold float64, releaseThreshold float64)
+}
+
+type CNSIPAMPoolMonitor struct {
+	initialized bool
+
+	cns *HTTPRestService
+	rc  requestcontroller.RequestController
+
+	batchSize        int
+	requestThreshold float64
+	releaseThreshold float64
+	minimumFreeIps   int
+	maximumFreeIps   int
+	sync.RWMutex
+}
+
+func NewCNSIPAMPoolMonitor(cnsService *HTTPRestService, requestController requestcontroller.RequestController) *CNSIPAMPoolMonitor {
+	return &CNSIPAMPoolMonitor{
+		initialized: false,
+		cns:         cnsService,
+		rc:          requestController,
+	}
+}
+
+// TODO: add looping and cancellation to this, and add to CNS MAIN
+func (pm *CNSIPAMPoolMonitor) Start() error {
+
+	if pm.initialized {
+		availableIPConfigs := pm.cns.GetAvailableIPConfigs()
+		rebatchAction := pm.checkForResize(len(availableIPConfigs))
+		switch rebatchAction {
+		case increasePoolSize:
+			return pm.increasePoolSize()
+		case decreasePoolSize:
+			return pm.decreasePoolSize()
+		}
+	}
+	return nil
+}
+
+// UpdatePoolLimitsTransacted called by request controller on reconcile to set the batch size limits
+func (pm *CNSIPAMPoolMonitor) UpdatePoolLimitsTransacted(batchSize int, requestThreshold float64, releaseThreshold float64) {
+
+	pm.Lock()
+	defer pm.Unlock()
+	pm.batchSize = batchSize
+	pm.requestThreshold = requestThreshold
+	pm.releaseThreshold = releaseThreshold
+
+	//TODO: rounding
+	pm.minimumFreeIps = int(float64(pm.batchSize) * requestThreshold)
+	pm.maximumFreeIps = int(float64(pm.batchSize) * releaseThreshold)
+	pm.initialized = true
+}
+
+func (pm *CNSIPAMPoolMonitor) checkForResize(freeIPConfigCount int) int {
+	switch {
+	// pod count is increasing
+	case freeIPConfigCount < pm.minimumFreeIps:
```

Comment:
You also need to account existing UpdatedCount, scenario we discussed in the meeting where SpecCount is already updated (so now freeIp = AvailableIPs + (If UpdatedSpecCount > TotalIPCount then (UpdatedSpec - TotalIpCount), else 0))